### PR TITLE
Hashing of assets

### PR DESCRIPTION
**Pull request by Hazardu**
##Description
Adds new methods to `IAsset` interface as a prerequisite for asset converter v2

## IAssets derived classes
A list of classes that need their implementation of `hash`, `compare`, `canBeRestoredFrom` written/changed
- [x] ICPUBufferView
- [x] ICPUAccelerationStructure
- [x] ICPUAnimationLibrary
- [x] ICPUBuffer
- [x] ICPUCommandBuffer
- [x] ICPUComputePipeline
- [x] ICPUDescriptorSet
- [x] ICPUDescriptorSetLayout
- [x] ICPUEvent
- [x] ICPURenderpass
- [x] ICPUFramebuffer
- [x] ICPUGraphicsPipeline
- [x] ICPUFramebuffer
- [x] ICPUImage
- [x] ICPUImageView
- [ ] ICPUMesh
- [ ] ICPUMeshBuffer
- [ ] ICPUPipelineCache
- [ ] ICPUPipelineLayout
- [ ] ICPUSpecializedShader
- [ ] ICPURenderpass
- [ ] ICPURenderpassIndependentPipeline
- [ ] ICPUSampler
- [ ] ICPUShader
- [ ] ICPUSkeleton ?

### CODE REVIEW

**Comment by Hazardu** [comment]
@devshgraphicsprogramming Could you take a look at the implementation of the new methods in `ICPUBufferView` before I make analogous methods in the other IAsset derived classes?

**Comment by Hazardu** [comment]
Also i have not seen any mention of `ICPUSkeleton` in #376 
Is this something obsolete?

**Comment by devshgraphicsprogramming** [comment]
> Also i have not seen any mention of `ICPUSkeleton` in #376 Is this something obsolete?

not fully developed yet

File: include/nbl/asset/IAsset.h [hunk 1]
```diff
@@ -210,6 +209,13 @@ class IAsset : virtual public core::IReferenceCounted
 
 		virtual bool canBeRestoredFrom(const IAsset* _other) const = 0;
 
+		//! 
+		virtual size_t hash() const = 0;
+
+		//!
+		virtual bool equals(const IAsset* _other) const = 0;
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/IAsset.h:216]: you want to run basic checks here (like `_other` being the same asset type, etc.) and use a `protected virtual equals_impl` for the details

you probably want to do a similar refactor to `canBeRestoredFrom`

File: include/nbl/asset/ICPUBufferView.h [hunk 2]
```diff
@@ -96,6 +105,12 @@ class ICPUBufferView : public IBufferView<ICPUBuffer>, public IAsset
 		}
 
 		virtual ~ICPUBufferView() = default;
+
+	private:
+
+		bool compareMembers(const ICPUBufferView* other) const {
+			return (m_size == other->m_size) && (m_offset == other->m_offset) && (m_format == other->m_format);
+		}
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUBufferView.h:113]: ok we want to move it up to `IAsset` and call it something like `protected: bool compatible`

then every asset's `canBeRestoredFrom` and `equals` can call into `compatible()` first

File: include/nbl/asset/ICPUCommandBuffer.h [hunk 3]
```diff
@@ -41,19 +41,42 @@ class ICPUCommandBuffer final :
...
-    bool canBeRestoredFrom(const IAsset* _other) const
-    {
-        return false; // TODO
-    }
-
+    _NBL_STATIC_INLINE_CONSTEXPR auto AssetType = ET_COMMAND_BUFFER;
     E_TYPE getAssetType() const override
     {
-        return ET_COMMAND_BUFFER;
+        return AssetType;
```
Comment (Hazardu) [inline include/nbl/asset/ICPUCommandBuffer.h:47]: Even though this entire class is incomplete, ive changed the getAssetType to look similar to other IAsset derived classes

**Review by Hazardu — COMMENTED** [review]
I have messed up some things by hashing immutable fields in classes without any other members

File: include/nbl/asset/ICPUAccelerationStructure.h [hunk 4]
```diff
@@ -242,22 +242,61 @@ class ICPUAccelerationStructure final : public IAccelerationStructure, public IA
...
 		}
 
-		//!
-		_NBL_STATIC_INLINE_CONSTEXPR auto AssetType = ET_ACCELERATION_STRUCTURE;
-		inline IAsset::E_TYPE getAssetType() const override { return AssetType; }
 
+		size_t hash(std::unordered_map<IAsset*, size_t>* temporary_hash_cache = nullptr) const override
+		{
+			size_t seed = AssetType;
+			core::hash_combine(seed, params);
```
[the remark this answers is not in the document]
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUAccelerationStructure.h:268]: wdym by immutable params?

[the remark this answers is not in the document]
Comment (Hazardu) [reply include/nbl/asset/ICPUAccelerationStructure.h:268]: creation params

File: include/nbl/asset/IAsset.h [hunk 5]
```diff
@@ -242,6 +248,26 @@ class IAsset : virtual public core::IReferenceCounted
 			//_NBL_DEBUG_BREAK_IF(imm);
 			return imm;
 		}
+		//! compares two objects for matching types and, in derived types, compares member variable values. Common code for IAsset::equals, IAsset::canBeRestoredFrom
+		virtual bool compatible(const IAsset* _other) const {
+			_other->getAssetType() == getAssetType();
+		}
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/IAsset.h:254]: follow this pattern
```cpp
inline bool compatible() const
{
    if (getAssetType()==_other->getAssetType())
        return false;
    return compatible_impl();
}
```

where `_impl` is a pure virtual method.

File: include/nbl/asset/ICPUAccelerationStructure.h [hunk 6]
```diff
@@ -242,22 +242,58 @@ class ICPUAccelerationStructure final : public IAccelerationStructure, public IA
...
+		bool equals(const IAsset* _other) const override
+		{
+			return compatible(_other);
+		}
+
+
 		bool canBeRestoredFrom(const IAsset* _other) const override
 		{
 			auto* other = static_cast<const ICPUAccelerationStructure*>(_other);
 			_NBL_TODO();
```
[the remark this answers is not in the document]
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUAccelerationStructure.h:260]: thats fine, as long as you make a `compatible_impl()` with an `_NBL_TODO`

File: include/nbl/asset/ICPUBuffer.h [hunk 7]
```diff
@@ -107,7 +99,37 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
...
+		}
+
     protected:
+
+	    bool compatible(const IAsset* _other) const override {
+		    if (IAsset::compatible(_other)) {
+			    auto* other = static_cast<const ICPUBuffer*>(_other);
+                return (m_creationParams.size == other->m_creationParams.size);
+		    }
+		    return false;
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUBuffer.h:130]: if you followed the `_impl` pattern, youu would not need to remember to call `IAsset::compatible` and you'd be sure that th function is called with compat asset already

File: include/nbl/asset/ICPUBuffer.h [hunk 8]
```diff
@@ -107,7 +99,37 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
             return true;
         }
 
+        bool equals(const IAsset* _other) const override
+		{
+			auto* other = static_cast<const ICPUBuffer*>(_other);
+			return compatible(other) && (m_creationParams.usage.value == other->m_creationParams.usage.value);
```
[the remark this answers is not in the document]
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUBuffer.h:105]: you want to compare usage in `compatible_impl` probably

and please compare pointers not contents, will be much much faster and more accurate

P.S. you also want to turn `equals` into `equals_impl` pattern too

File: include/nbl/asset/IAsset.h (hunk 5 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/IAsset.h:254]: same thing for `equals`

```cpp
inline bool equals(const IAsset* _other) const
{
   if (compatible(_other))
        return equals_impl(_other);
   return false;
}
```

File: include/nbl/asset/ICPUBuffer.h [hunk 9]
```diff
@@ -107,7 +99,37 @@ class ICPUBuffer : public asset::IBuffer, public asset::IAsset
...
+			auto* other = static_cast<const ICPUBuffer*>(_other);
+			return compatible(other) && (m_creationParams.usage.value == other->m_creationParams.usage.value);
+		}
+
+
+		bool canBeRestoredFrom(const IAsset* _other) const override
+		{
+			auto* other = static_cast<const ICPUBuffer*>(_other);
+			return compatible(other);
+		}
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUBuffer.h:113]: can we just get rid of this, and replace its usages with `compatible` ?

File: include/nbl/asset/ICPUCommandBuffer.h (hunk 3 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUCommandBuffer.h:47]: we can probably nuke that entire class/header, we intend to have a GPU-Task-Graph instead to deal with synchronisation/layout transitions later

**Comment by devshgraphicsprogramming** [comment]
btw can you make `IAsset()` constructor protected?

File: include/nbl/asset/ICPUBuffer.h (hunk 9 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUBuffer.h:113]: hmm probably not the case for other higher level assets

**Comment by devshgraphicsprogramming** [comment]
Another thing is if we could have the `clone` follow the `_impl` pattern, so instead of the overriden `clone` needing to remember to call `clone_common`, we'd rename the current `public core::Smart_refctd_ptr<IAsset> clone`, into `protected core::smart_refctd_ptr<IAsset> clone_impl`


and exchange this
```cpp
public: 
		//! creates a copy of the asset, duplicating dependant resources up to a certain depth (default duplicate everything)
        virtual core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const = 0;
protected:
        inline void clone_common(IAsset* _clone) const
        {
            assert(!isDummyObjectForCacheAliasing);
            _clone->isDummyObjectForCacheAliasing = false;
            _clone->m_mutability = EM_MUTABLE;
        }
```
for
```cpp
public: 
        //! creates a copy of the asset, duplicating dependant resources up to a certain depth (default duplicate everything)
        inline core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const
        {
                assert(!isDummyObjectForCacheAliasing);
                auto clone = clone_impl(_depth);
                _clone->isDummyObjectForCacheAliasing = false;
                _clone->m_mutability = EM_MUTABLE;
                return clone;
        }
protected:
        virtual core::smart_refctd_ptr<IAsset> clone_impl(uint32_t _depth) const = 0;
```

**Comment by devshgraphicsprogramming** [comment]
> Another thing is if we could have the `clone` follow the `_impl` pattern, so instead of the overriden `clone` needing to remember to call `clone_common`, we'd rename the current `public core::Smart_refctd_ptr<IAsset> clone`, into `protected core::smart_refctd_ptr<IAsset> clone_impl`
> 
> and exchange this
> 
> ```c++
> public: 
> 		//! creates a copy of the asset, duplicating dependant resources up to a certain depth (default duplicate everything)
>         virtual core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const = 0;
> protected:
>         inline void clone_common(IAsset* _clone) const
>         {
>             assert(!isDummyObjectForCacheAliasing);
>             _clone->isDummyObjectForCacheAliasing = false;
>             _clone->m_mutability = EM_MUTABLE;
>         }
> ```
> 
> for
> 
> ```c++
> public: 
>         //! creates a copy of the asset, duplicating dependant resources up to a certain depth (default duplicate everything)
>         inline core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const
>         {
>                 assert(!isDummyObjectForCacheAliasing);
>                 auto clone = clone_impl(_depth);
>                 _clone->isDummyObjectForCacheAliasing = false;
>                 _clone->m_mutability = EM_MUTABLE;
>                 return clone;
>         }
> protected:
>         virtual core::smart_refctd_ptr<IAsset> clone_impl(uint32_t _depth) const = 0;
> ```

Also do a similar pattern for `converToDummyObject`:
```cpp
public:
         inline void convertToDummyObject(uint32_t referenceLevelsBelowToConvert=0u)
         {
                 if (canBeConvertedToDummy())
                         isDummyObjectForCacheAliasing = true;
                 convertToDummyObject_impl();
         }
protected:
         virtual void convertToDummyObject_impl(uint32_t referenceLevelsBelowToConvert) = 0;
```

Then everything is easier to read than the old sphaggetti.

File: include/nbl/asset/ICPUBufferView.h [hunk 10]
```diff
@@ -64,38 +65,56 @@ class ICPUBufferView : public IBufferView<ICPUBuffer>, public IAsset
...
 
-	protected:
-		void restoreFromDummy_impl(IAsset* _other, uint32_t _levelsBelow) override
+
+		bool canBeRestoredFrom(const IAsset* _other) const override
 		{
-			auto* other = static_cast<ICPUBufferView*>(_other);
+			auto* other = static_cast<const ICPUBufferView*>(_other);
+			return compatible(other) && m_buffer->canBeRestoredFrom(other->m_buffer.get());
+		}
```
Comment (devshgraphicsprogramming) [inline include/nbl/asset/ICPUBufferView.h:79]: https://discord.com/channels/593902898015109131/1078226704092381194/1083062053939445821

**Comment by devshgraphicsprogramming** [comment]
Also `restoreFromDummy_impl_call` needs to be templated on `class AssetT` cause otherwise its not doing anything useful for us

File: include/nbl/asset/ICPUBufferView.h (hunk 10 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUBufferView.h:79]: Superceded by https://discord.com/channels/593902898015109131/1078226704092381194/1083069244650045600

File: include/nbl/asset/ICPUBuffer.h (hunk 9 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUBuffer.h:113]: https://discord.com/channels/593902898015109131/1078226704092381194/1083069244650045600

File: include/nbl/asset/ICPUBuffer.h (hunk 7 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/ICPUBuffer.h:130]: https://discord.com/channels/593902898015109131/1078226704092381194/1083069244650045600

File: include/nbl/asset/IAsset.h (hunk 5 above)
Comment (devshgraphicsprogramming) [reply include/nbl/asset/IAsset.h:254]: https://discord.com/channels/593902898015109131/1078226704092381194/1083069244650045600

**Pull request closed without merging**